destination/destination_test.go

```diff
@@ -0,0 +1,256 @@
+		is := is.New(t)
+
+		d := NewDestination()
+
+		ctx := context.Background()
+
+		err := d.Configure(ctx, map[string]string{
+			ConfigDatabase: "projects/project_id/instances/instance_id/databases/database_id",
+			ConfigEndpoint: "localhost:9010",
+		})
+		is.NoErr(err)
+
+		err = d.Open(ctx)
+		is.NoErr(err)
+
+		d.Teardown(ctx)
+	})
+
+	t.Run("destination open failure", func(t *testing.T) {
+		is := is.New(t)
+
+		d := NewDestination()
+
+		ctx := context.Background()
+
+		err := d.Configure(ctx, map[string]string{
+			ConfigDatabase: "projects/project_id/instances/instance_id/databases",
+			ConfigEndpoint: "localhost:9010",
+			ConfigTable:    `{{ index .Metadata }}`,
+		})
+		is.NoErr(err)
+
+		err = d.Open(ctx)
+		is.True(err != nil)
+		is.Equal(err.Error(), `failed to create spanner client: failed to create spanner client: database name "projects/project_id/instances/instance_id/databases" should conform to pattern "^projects/(?P<project>[^/]+)/instances/(?P<instance>[^/]+)/databases/(?P<database>[^/]+)$"`)
+
+		d.Teardown(ctx)
+	})
+}
+
+func TestDestination_Write(t *testing.T) {
+	t.Parallel()
+
+	t.Run("destination write insert success", func(t *testing.T) {
+		is := is.New(t)
+		ctx := context.Background()
+		testutils.SetupDatabase(ctx, is)
+
+		d := NewDestination()
+
+		err := d.Configure(ctx, map[string]string{
+			ConfigDatabase: fmt.Sprintf("projects/%s/instances/%s/databases/%s", testutils.ProjectID, testutils.InstanceID, testutils.DatabaseID),
+			ConfigEndpoint: "localhost:9010",
+		})
+		is.NoErr(err)
+
+		err = d.Open(ctx)
+		is.NoErr(err)
+
+		records := []opencdc.Record{
+			{
+				Operation: opencdc.OperationCreate,
+				Metadata:  opencdc.Metadata{"opencdc.collection": "Singers"},
+				Key:       opencdc.StructuredData{"SingerID": 1},
+				Payload: opencdc.Change{
+					After: opencdc.StructuredData{
+						"SingerID":  "1",
+						"Name":      "jon",
+						"CreatedAt": time.Now().UTC(),
+					},
+				},
+			},
+		}
+
+		n, err := d.Write(ctx, records)
+		is.NoErr(err)
+		is.Equal(n, len(records))
+
+		d.Teardown(ctx)
+	})
+
+	t.Run("destination write update success", func(t *testing.T) {
+		is := is.New(t)
+		ctx := context.Background()
+		d := NewDestination()
+
+		err := d.Configure(ctx, map[string]string{
+			ConfigDatabase: fmt.Sprintf("projects/%s/instances/%s/databases/%s", testutils.ProjectID, testutils.InstanceID, testutils.DatabaseID),
+			ConfigEndpoint: "localhost:9010",
+		})
+		is.NoErr(err)
+
+		err = d.Open(ctx)
+		is.NoErr(err)
+
+		records := []opencdc.Record{
+			{
+				Operation: opencdc.OperationUpdate,
+				Metadata:  opencdc.Metadata{"opencdc.collection": "Singers"},
+				Key:       opencdc.StructuredData{"SingerID": 1},
+				Payload: opencdc.Change{
+					After: opencdc.StructuredData{
+						"SingerID":  "1",
+						"Name":      "jon",
+						"CreatedAt": time.Now().UTC(),
+					},
+				},
+			},
+		}
+
+		n, err := d.Write(ctx, records)
+		is.NoErr(err)
+		is.Equal(n, len(records))
```

Comment:
can we add some kind of assertion that the written record equals the one in `records`, at least asserting the key value

Reply:
We have to query the record and then compare.